### ISSUE TITLE
Potential race condition in Free calls between Remove and Dispose

### ISSUE THREAD

**Issue by ProductiveRage** [pvginkel/InMemCached#1]
I'm trying to review the code to try to ensure that there are no gremlins hiding in the multi-threaded complexity.. and I'm finding it quite difficult! Let me know if any of what I've written here seems incorrect..

In the "Add" method, you're not locking the handle when the _cache.TryAdd call fails since no other threads are able to get a reference to that handle and so there will never be multiple threads trying to manipulate it. This is unlike in "Get" and "Remove" where locking _is_ required because it's possible for one thread to enter the "_cache.TryGetValue" in "Get" condition and for another to enter the "_cache.TryRemove" condition in "Remove" simultaneously; in this case, only one of these operations may be allowed at a time (and if "Remove" gets there first then "Get" will have to return null since the handle has been free'd, which is why "ToByteArray" may return null).

I presume that you're applying similar logic to the lock in the "Dispose" method. However, if that is the case then there is a potential race condition. The "Values" property on the ConcurrentDictionary is a snapshot of Values that are present at the point at which the property is accessed. If any values are removed from the dictionary during the enumeration, this will not affect the values that are enumerated over. For example, if there are three handles to enumerate over and the third handle is removed by a "Remove" call on a thread that executes while another thread is already enumerating the dictionary values from within "Dispose" then "Dispose" will try to Free the handle that "Remove" has already free'd.

This is illustrated with the below example:

```
class Program
{
    static void Main(string[] args)
    {
        var cache = new ConcurrentDictionary<int, MemoryHandle>();
        cache.TryAdd(1, new MemoryHandle());
        new Thread(() =>
        {
            foreach (var value in cache.Values)
            {
                // Introduce a short delay to ensure that the TryRemove call below completes
                // before this enumeration completes
                Thread.Sleep(TimeSpan.FromSeconds(1));
                value.Free();
            }
        }).Start();
        new Thread(() =>
        {
            MemoryHandle removedHandle;
            if (cache.TryRemove(1, out removedHandle))
                removedHandle.Free();
        }).Start();
        Console.ReadLine();
    }

    private sealed class MemoryHandle
    {
        private bool _freed = false;
        public void Free()
        {
            if (_freed)
                throw new Exception("This handle has already been free'd!");
            _freed = true;
        }
    }
}
```

Maybe it would be best to change "Free" to work like "ToByteArray" - in that it will perform no work if the handle is a null pointer?

**Comment by pvginkel** [comment]
I'm not sure if I understand. So, the `Free` is only called within the scope of a `lock`, the same one that `ToByteArray` is called in, so there is no race there. This also applies to the `Dispose` of the `InMemCached` class, so that should not be an issue.

The only place where I don't lock is when clearing the `ConcurrentDictionary<,>`, but that's not an issue since the `Dispose` of the `InMemCached` assumes that it is not in use anymore. Even this is not a problem since the `Free` itself in the `Dispose` is surrounded by a `lock`.

The code sample you gave uses a `ConcurrentDictionary<,>` directly, and I understand the issue your describing. However, I'm not sure how this translates to the `InMemCached` class, since I don't think this issue applies.

Btw, the benchmark project that's part of the repository does implement multi threading. There are asserts throughout the code so running that benchmark in debug mode should give you a good picture of any issues that may arise. Specifically, translating to your code sample, the `Free` has an assert on whether `_handle` is zero and if so, raises the assert.

If you can cause a threading issue, maybe with changes to the benchmark code, please let me know.

**Comment by ProductiveRage** [comment]
I can't immediately see a way to create a threading issue so long as we are happy with the assumption that Dispose should only be called when the InMemCached instance is no longer to be used (which I think is a reasonable assumption).

So I guess that the question, really, is why do you have a lock in the Dispose method? It would not be valid for multiple threads to call Dispose and (based upon the above assumption) it would not be valid for any other calls to occur concurrently with a Dispose call - so the lock isn't required?

**Comment by pvginkel** [comment]
Good question, I don't have an answer. Do you think it should be removed?

**Comment by ProductiveRage** [comment]
Yes, I do. It's great that you've covered threading scenarios in the Benchmark project, I was just trying to mentally reason about it myself _as well_.

The next area that I'll be concentrating on is GetMemoryHandle / ReleaseMemoryHandle. It looks like the approach that you've used is very similar to as what the C# compiler generates for the add / remove logic for event handlers? Which means that it's almost certainly correct, I'm just trying to convince myself that it is and that it's safe to set "result.next" in the way that you are doing. If I think that I find anything wrong then I'll raise a separate issue. This multi-threading malarkey is hard (or maybe it's just me! :)

**Comment by pvginkel** [comment]
Ok, I'll have a look at that. The Interlocked mechanism actually is a quite
standard one, but this stuff is almost impossible to reason about. I'd
suggest finding documentation and a sample implementation from a source you
trust and verify it follows that pattern. That's what I did.
